corehq/apps/geospatial/static/geospatial/js/case_grouping_map.js

```diff
@@ -328,10 +325,102 @@ hqDefine("geospatial/js/case_grouping_map",[
             (c ^ crypto.getRandomValues(new Uint8Array(1))[0] & 15 >> c / 4).toString(16),
         );
     }
+
     function caseGroupSelectModel() {
         'use strict';
         var self = {};
 
+        self.allGroups = ko.observableArray([]);
+        self.allCaseGroups;
+        self.showGroupSelectionTable = ko.computed(function() {
+            return groupLockModelInstance.groupsLocked();
+        })
+        self.visibleGroupIDs = ko.observableArray([]);
+
+        self.groupIDInVisibleGroupIds = function(groupID) {
+            if (self.visibleGroupIDs().indexOf(groupID) !== -1) {
+                return true;
+              } else {
+                return false;
+              }
+        };
+
+        self.getGroupByID = function(groupID) {
+            return _.filter(self.allGroups(), function(group) {return group.groupID === groupID})[0];
+        };
+
+        self.loadCaseGroups = function(caseGroups) {
+            self.allCaseGroups = caseGroups; // {case_id: {group_info}}
+            // Add groups to the cases being exported
+            exportModelInstance.addGroupsToCases(caseGroups);
+            let uniqueGroups = [];
+            var groupIds = _.map(caseGroups, function(caseItem) {
+                return caseItem.groupId;
+            });
+            new Set(groupIds).forEach(id => uniqueGroups.push(
+                {groupID: id, color: getRandomRGBColor()}
+            ));
+            self.allGroups(uniqueGroups);
+
+            let visibleIDs = _.map(uniqueGroups, function(group) {return group.groupID});
+            self.visibleGroupIDs(visibleIDs);
+        };
+
+        self.restoreMarkerOpacity = function() {
+            mapMarkers.forEach(function(marker) {
+                setMarkerOpacity(marker, DEFAULT_MARKER_OPACITY);
+            });
+        };
+
+        self.highlightGroup = function(group) {
+            exportModelInstance.casesToExport().forEach(caseItem => {
+                    let caseIsInGroup = caseItem.groupId === group.groupID;
+                    mapMarkers.forEach(function(marker) {
```

Comment:
Will there be more than one marker for each `caseItem`? If not, we could rather a `mapMarkers.find()` instead of a `forEach()` to prevent looping through the rest of the markers after we've found a match.

Reply:
.find seems better yeah

Reply:
https://github.com/dimagi/commcare-hq/pull/33680/commits/b514f9e1390c71bc57c0e914b872458f0532afa6